### PR TITLE
test: more complex input queues scenario for queue compatibility tests

### PR DESCRIPTION
**Pull request by oggy-dfin**
Add a more exhaustive test to check that the visible behavior of canister input queues stays stable, in particular:
1. that we pick input messages in a round-robin fashion over the different queue types
2. that we successfully drain all messages from input queues

### CODE REVIEW

File: rs/sns/root/src/lib.rs [hunk 1]
```diff
@@ -1685,7 +1685,7 @@ mod tests {
     async fn register_dapp_canisters_redundant() {
         // Step 1: Prepare the world.
         thread_local! {
-            static DAPP_CANISTER_ID: PrincipalId = PrincipalId::new_user_test_id(4);
+            static DAPP_CANISTER_ID: PrincipalId = const { PrincipalId::new_user_test_id(4) };
```
Comment (oggy-dfin) [inline rs/sns/root/src/lib.rs:1688]: The change here is due to Clippy

**Review by alin-at-dfinity — APPROVED** [review]

File: rs/types/base_types/src/principal_id.rs [hunk 2]
```diff
@@ -196,11 +196,23 @@ impl PrincipalId {
...
+    pub const fn new_user_test_id(n: u64) -> Self {
+        let mut bytes = [0u8; Self::MAX_LENGTH_IN_BYTES];
+        let n_bytes = n.to_le_bytes();
+
+        // Copy the u64 bytes into the array
+        let mut i = 0;
+        while i < n_bytes.len() {
+            bytes[i] = n_bytes[i];
+            i += 1;
+        }
```
[the remark this answers is not in the document]
Comment (oggy-dfin) [reply rs/types/base_types/src/principal_id.rs:208]: Nope, apparently Rust doesn't allow for loops in const fns:
https://github.com/rust-lang/rust/issues/87575

I should've left a comment though, added one now.

File: rs/replicated_state/src/canister_state/queues/tests.rs [hunk 3]
```diff
@@ -2348,4 +2349,92 @@ mod mainnet_compatibility_tests {
...
+
+            // Put a request and a response from a local canister in the input queues
+            queues
+                .push_input_request(LOCAL_CANISTER_ID, InputQueueType::LocalSubnet)
+                .unwrap();
+            // Make an outgoing request just to create a reservation for a response
+            queues.push_output_request(LOCAL_CANISTER_ID).unwrap();
+            queues
+                .push_input_response(LOCAL_CANISTER_ID, InputQueueType::LocalSubnet)
+                .unwrap();
```
[the remark this answers is not in the document]
Comment (oggy-dfin) [reply rs/replicated_state/src/canister_state/queues/tests.rs:2379]: Makes sense, added a helper.

File: rs/tests/message_routing/queues_compatibility_test.rs [hunk 4]
```diff
@@ -140,27 +145,66 @@ fn test_one_direction(
...
+
+    pub fn run(&self, mainnet_version: &str, logger: &Logger) {
+        match &self.test_type {
+            TestType::Bidirectional { published_binary } => {
+                self.self_test(logger);
+                self.bidirectional_test(mainnet_version, published_binary, logger)
+            }
+            TestType::SelfTestOnly => {
+                self.self_test(logger);
+            }
```
[the remark this answers is not in the document]
Comment (oggy-dfin) [reply rs/tests/message_routing/queues_compatibility_test.rs:176]: Yep, changed now, thanks!

**Review by daniel-wong-dfinity-org — APPROVED** [review]

**Review by marko-k0 — APPROVED** [review]

**Pull request merged**